File: 1. 단계별로_풀어보기/20~29/24. 분할 정복/4. 1629.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [A, B, C] = inputArr[0].split(' ').map(BigInt)

// 거듭제곱별 나머지 저장하는 배열
const modArr = Array.from({length: 33}).fill(0n)
modArr[0] = A % C


// 거듭제곱별 나머지 배열 채우기
for (let i = 1; i < 33; i++) {
  modArr[i] = (modArr[i - 1] * modArr[i - 1]) % C
}
 
let result = 1n

let remain = B
let idx = 0

while (remain > 0n) {
  // 가장 낮은 비트갑을 구한다.
  const mul = remain & 1n
  
  if (mul) {
    result = (result * modArr[idx]) % C
  }

  remain = remain >> 1n
  idx++
}

process.stdout.write(`${result}`)
 

File: 1. 단계별로_풀어보기/20~29/24. 분할 정복/4. 1629.py
import sys

input = sys.stdin.readline
output = sys.stdout.write

A, B, C = map(int, input().split())

modArr = [0] * 33
modArr[0] = A % C

for i in range(1, 33):
  modArr[i] = (modArr[i - 1] * modArr[i - 1]) % C

result = 1

remain = B
idx = 0

while remain > 0:
  mul = remain & 1
  
  if mul:
    result = (result * modArr[idx]) % C

  remain = remain // 2
  idx += 1

output(f"{result}")